src/lib/block-batch.js

```diff
@@ -1,35 +1,42 @@
 const MAX_BYTES_BETWEEN = 1024 * 1024 * 2
 const MAX_BATCH_SIZE = 10
 
-export class BlockBatch {
-  /** @type {number[]} */
-  #offsets = []
+/**
+ * @typedef {import('multiformats').CID} CID
+ * @typedef {{ carCid: CID, blockCid: CID, offset: number }} BatchItem
+ * @typedef {{ add: (i: BatchItem) => void, next: () => BatchItem[] }} BlockBatcher
+ */
 
-  /**
-   * Add an offset to the batch
-   * @param {number} offset
-   */
-  add (offset) {
-    this.#offsets.push(offset)
+/**
+ * Batcher for blocks in CARs. Batches are grouped by CAR CID and blocks are
+ * returned in batches in the order they were inserted.
+ * @implements {BlockBatcher}
+ */
+export class OrderedCarBlockBatcher {
+  /** @type {BatchItem[]} */
+  #queue = []
+
+  /** @param {BatchItem} item */
+  add (item) {
+    this.#queue.push(item)
   }
 
   next () {
-    const offsets = this.#offsets // .sort((a, b) => a - b)
-    if (!offsets.length) return
+    const queue = this.#queue
+    if (!queue.length) return []
     const batch = []
-    let last = offsets[0]
+    let last = queue[0]
```

Comment:
last is first item in queue? i guess it's the last from the previous batch, but it's a little confusing that we take the head of the queue and call it last.

Reply:
also why not shift this guy off the queue and add him to the batch immediately, otherwise our first comparison is always with itself!